src/ebonite/core/objects/wrapper.py

```diff
@@ -28,52 +35,126 @@ class ModelWrapper(EboniteParams):
+        if input_data is None:
+            raise ValueError("Input data sample should be specified as 'input_data' key in order to analyze model")
+
         self.model = model
+        self.methods = self._prepare_methods(input_data)
         return self
 
+    def _prepare_methods(self, input_data):
+        arg_type = DatasetAnalyzer.analyze(input_data)
+        methods = {}
+        for exposed, wrapped in self._exposed_methods_mapping().items():
+            output_data = self._call_method(wrapped, input_data)
+            out_type = DatasetAnalyzer.analyze(output_data)
+            methods[exposed] = (wrapped, arg_type, out_type)
+        return methods
+
     def unbind(self):
         """
         Unbind model object from this wrapper
         :return: self
         """
         self.model = None
+        self.methods = None
         return self
 
+    @property
+    def exposed_methods(self) -> typing.Set[str]:
+        if self.methods is None:
+            raise ValueError('Wrapper {} has no model yet'.format(self))
+        return set(self.methods.keys())
+
+    def method_signature(self, name) -> typing.Tuple[DatasetType, DatasetType]:
+        """
+        Determines input / output types of model wrapper method with given name
+
+        :param name: name of the method to determine input / output types
+        :return: input / output type of method with given name
+        """
+        self._check_method(name)
+        _, *signature = self.methods[name]
+        return signature
+
+    def call_method(self, name, input_data):
+        """
+        Calls model wrapper method with given name on given input data
+
+        :param name: name of the method to call
+        :param input_data: argument for the method
+        :return: call result
+        """
+        self._check_method(name)
+        wrapped, *_ = self.methods[name]
+        output_data = self._call_method(wrapped, input_data)
+        return output_data
+
+    def _check_method(self, name):
+        if self.model is None:
+            raise ValueError('Wrapper {} has no model yet'.format(self))
+        if name not in self.methods:
+            raise ValueError(f"Wrapper '{self}' obj doesn't expose method '{name}'")
+
+    def _call_method(self, wrapped, input_data):
+        if wrapped is None:
+            return self.model(input_data)
+        if hasattr(self, wrapped):
+            return getattr(self, wrapped)(input_data)
+        return getattr(self.model, wrapped)(input_data)
+
     @abstractmethod
-    def predict(self, data):
+    def _exposed_methods_mapping(self) -> typing.Dict[str, typing.Optional[str]]:
         """
-        Must return model prediction on data
+        Should return methods exposed by this model wrapper
 
-        :param data: data to predict
-        :return: prediction
+        :return: methods dict: exposed method name to wrapped method name
+        If wrapped method name is `None` then wrapped model object should be called directly.
```

Comment:
maybe just use `__call__` instead of None?

Reply:
fixed